hpx/lcos/detail/future_data.hpp

```diff
@@ -489,10 +496,20 @@ namespace detail
         /// operation. Allows any subsequent set_data operation to succeed.
         void reset(error_code& /*ec*/ = throws)
         {
-            boost::unique_lock<mutex_type> l(this->mtx_);
+            state s = empty;
+
+            {
+                boost::unique_lock<mutex_type> l(this->mtx_);
+
+                std::swap(state_, s);
+                on_completed_ = completed_callback_type();
+
+                // unlock for the destruction of embedded data
+                // FIXME: is that correct?
+            }
```

Comment:
This change is unnecessary. A `reset` has to be non-observable, or futures would be unable to maintain their invariants. It can only be performed by a single writer with no readers.